kucoin backend

jaredgoldman opened the pull request:
This PR includes a brokerd module for [Kucoin](https://www.kucoin.com)'s [api](https://www.kucoin.com/api). So far, only code for market data has been implemented. This includes:

- Fetching a users historical trades
- Subscribing to real-time market data via the Kucoin websocket feed
- Multi-chartz support
- Fetching OHLC historical market data
- Kucoin user auth
- Symbol fuzzy search
---
Remaining tasks for full support are detailed in [this issue](https://github.com/pikers/piker/issues/499)

goodboy commented:
@jaredgoldman this ready for serious review?

jaredgoldman commented:
@goodboy I think we're ready for some more testing

goodboy commented:
@algorandpa seems like i can only stage order with `size=0` but other then that historical and live data seems right 🏄🏼

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+'''
+
+from random import randint
+from typing import Any, Callable, Optional, Literal, AsyncGenerator
```
as per other code,i prefer the multi-line version:

```python
from typing import (
	Any,
	Callable,
	Optional,
	Literal,
	AsyncGenerator,
)
```

jaredgoldman replied:
Done

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+
+log = get_logger(__name__)
+
+_ohlc_dtype = [
```
don't think you need to define this unless it's diff from our default.

goodboy replied:
yah, but i think if you look, the default schema def is applied if not defined..

it should only be set if something is diff from the default iirc

we should probably document this.

jaredgoldman replied:
This confuses me. I'm using it to parse my data into the numpy array
```py
            for _, (field_name, field_type) in enumerate(_ohlc_dtype):
                value = data[field_name]

                match field_name:
                    case 'index':
                        row.append(int(value))
                    case 'time':
                        row.append(value)
                    case _:
                        row.append(float(value))

            new_bars.append(tuple(row))
```

goodboy replied:
sure moreso my point is i don't think you need to define it at all in the module unless it's diff then the default:
https://github.com/pikers/piker/blob/master/piker/data/_source.py#L34

the logic which decides i believe is here in the data layer:
https://github.com/pikers/piker/blob/master/piker/data/feed.py#L799

so i'm just saying you might not need to even define it and if you need to iterate fields, just iterate the `base_iohlc_dtype`?

further this whole loop thinger:
```python
  for _, (field_name, field_type) in enumerate(_ohlc_dtype):
                value = data[field_name]
```

is this really any better then just doing inline static assignment?

for eg. here's how we do it in the `ib` backend:
https://github.com/pikers/piker/blob/master/piker/brokers/ib/api.py#L271

you'll notice it's just based on ordering the fields according to the client lib/API's output format, and we just kinda inline map it to the default ohlc dtype.

jaredgoldman replied:
Word, I'll simplify the logic

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+    Kucoin's pair format
+
+    '''
+
```
i tend to *not* prefer having an extra line between `'''` and docstring since we already put one as last content line in doc string.

jaredgoldman replied:
Done

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+]
+
+
+class KucoinMktPair(Struct, frozen=True):
```
put a comment with a link to api docs for this.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+    symbol: str
+
+
+class AccountTrade(Struct, frozen=True):
```
Again, links to API docs for all these structs.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+    def __init__(self) -> None:
+        self._pairs: dict[str, KucoinMktPair] = {}
+        self._bars: list[list[float]] = []
+        self._key_id: str
```
are these duplicate with the fields of `BrokerConfig`?
If so why do you need both sets?

goodboy replied:
just do `self._config: BrokerConfig` no?

fwiw `msgspec.Struct`s don't do any validation, so packing it in one without doing anything else is basically a noop.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+
+        config: BrokerConfig | None = get_config()
+
+        if config and config.key_id and config.key_secret and config.key_passphrase:
```
as per above. either choose to wrap creds in a config var or do it in init method here; you don't need both.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+        Generic request wrapper for Kucoin API
+
+        '''
+        if self._authenticated:
```
shouldn't this be the inverse?

only regen auth headers when *not* already authed?

jaredgoldman replied:
Let me test this - I assumed since the headers need a timestamp that you'd have to create new ones for each request. In retrospect I could totally be wrong about this and haven't tested to evaluate my assumptions.

jaredgoldman replied:
Ah but wait, we have to pass the HTTP action into the header... @goodboy we have to encode to action and timestamp into the headers like so:
```py
str_to_sign = str(int(time.time() * 1000)) + action + f'/api/{api_v}{endpoint}'

        signature = base64.b64encode(
            hmac.new(
                self._key_secret.encode('utf-8'),
                str_to_sign.encode('utf-8'),
                hashlib.sha256,
            ).digest()
        )
```
Worth it to try to generate some of the headers ahead of time?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+
+    async def _get_ws_token(self, private: bool = False) -> tuple[str, int] | None:
+        '''
+        Fetch ws token needed for sub access
```
api docs link

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+        if data and 'token' in data:
+            ping_interval: int = data['instanceServers'][0]['pingInterval']
+            return data['token'], ping_interval
+        elif data:
```
what's normally happening in this case?

goodboy replied:
so if `bool(data)` being false-y should never happen yah?

cuz in that case this will exit silently i think?

jaredgoldman replied:
Let's account for that just in case. If the API request is malformed this could happen.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+    async def _get_pairs(
+        self,
+    ) -> dict[str, KucoinMktPair]:
+        if self._pairs:
```
wait, why?

jaredgoldman replied:
Whoops.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+
+    async def cache_pairs(
+        self,
+        normalize: bool = True,
```
no, do not mutate state and store is based on a flag 😂

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+            self._pairs = self._normalize_pairs(self._pairs)
+        return self._pairs
+
+    def _normalize_pairs(
```
you don't need both of these methods. just pick the one with the the normalized output, unless there's some reason not to?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+        return {kucoin_sym_to_fqsn(item[0].name): item[0] for item in matches}
+
+    async def last_trades(self, sym: str) -> list[AccountTrade]:
+        trades = await self._request('GET', f'/accounts/ledgers?currency={sym}', 'v1')
```
link to api docs.

goodboy commented:
@jaredgoldman as mentioned in chat as well we've got some kinda strange thing going on with trade rate calcs:
- there's some kind of trade event being picked up by the fsp system 10/sec
- you can see flat curve for this if you leave it running
- this **does not match up with** actual vlm / clearing rates

![screenshot-2023-03-29_21-32-25](https://user-images.githubusercontent.com/291685/229307708-d4abdc20-f663-4797-995f-17b11a8f9a2d.png)

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+
+def make_sub(sym, connect_id, level='l1') -> dict[str, str | bool]:
+    match level:
+        case 'l1':
```
Yeah so re my comment about the trade rates mis-calculation, I think this might be the cause.

maybe one of these subs is being reported every 10ms and you're not filtering / diffing for actual updates since the last msg?

likely you'll need to keep a local book state and only transmit to piker when there's a real update to the book and/or real clear events that have taken place.

goodboy replied:
yeah comment is just above https://github.com/pikers/piker/pull/494#issuecomment-1493068034

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+                sym: {
+                    'symbol_info': {
+                        'asset_type': 'crypto',
+                        'price_tick_size': 0.0005,
```
wait why are you hardcoding this instead of pulling it from the pair info in `pairs`?

jaredgoldman replied:
I'm actually unsure of what the `price_tick_size` and `lot_tick_size` are. Do they map with any values in `KucoinMktPair`?

jaredgoldman replied:
@goodboy - I think we need better documentation on init message formatting.

goodboy replied:
the short answer is yes but also this is being more formalized in #489 .. unfortunately a bit later then i would have liked 😂 

if you wanted to get ambitious you could try to adopt the stuff in there, but it's probably fine to get changed later.

more or less `price/lot_tick_size` are the float/Decimal values which specify the *min increment value* for both the market's price and volume/size.

goodboy replied:
so yes, you should pull them from `KucoinMktPair`.

likely the two values inline here:
https://github.com/pikers/piker/pull/494/files#diff-0e687c4074ef81501ccba2717591ca55647c53b9b3dc85c33d4b4fc1436c644cR78

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+                async def open_ping_task(ws: wsproto.WSConnection):
+                    '''
+                    Spawn a non-blocking task that pings the ws
+                    server every ping_interval so Kucoin doesn't drop
```
what's `ping_interval`?

jaredgoldman replied:
check out the docs in the docstring for `_get_ws_token`

jaredgoldman commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,668 @@
+        '''
+        now = int(time.time() * 1000)
+        path = f'/api/{api_v}{endpoint}'
+        str_to_sign = str(now) + action + path
```
slick

jaredgoldman commented:
@goodboy merged the binance changes and add anext()

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+        log.warning('No config section found for kucoin in config')
+        return None
+
+    return BrokerConfig(**section)
```
i kinda like this. if you wanted to ensure the typing you could call `.typecast()` i think?

https://github.com/pikers/piker/blob/master/piker/data/types.py#L83

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+    section = conf.get('kucoin')
+
+    if section is None:
```
what happens in this case?

jaredgoldman replied:
We do not generate auth headers for requests 
```py
 if self._config:
            headers = self._gen_auth_req_headers(action, endpoint, api_v)
```

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+        self,
+    ) -> dict[str, KucoinMktPair]:
+        entries = await self._request('GET', '/symbols')
+        syms = {kucoin_sym_to_fqsn(item['name']): KucoinMktPair(**item) for item in entries}
```
if you want you can be all hip and go with the new `fqme` from #489 😂

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+        entries = await self._request('GET', '/symbols')
+        syms = {kucoin_sym_to_fqsn(item['name']): KucoinMktPair(**item) for item in entries}
+
+        log.info('Kucoin market pairs fetched')
```
maybe toss the lenth of `syms` in this msg?

probably something we should do on all backends too

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+    async def cache_pairs(
+        self,
+        # normalize: bool = True,
```
drop this?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+        '''
+        if not self._pairs:
+            self._pairs = await self._get_pairs()
```
i think it'd be cool if in the longer run we actually cache these somehow to local storage per backend..

just thinking out loud.

jaredgoldman replied:
Great idea! Also a good opportunity for maybe a redis service that will periodically refresh it's cache.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+        pattern: str,
+        limit: int = 30,
+    ) -> dict[str, KucoinMktPair]:
+        data = await self._get_pairs()
```
wouldn't it make more sense to use `.cache_pairs()` for this call so you aren't constantly reloading?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+    async def _get_bars(
+        self,
+        fqsn: str,
+        start_dt: Optional[datetime] = None,
```
much prefer the new `datetime | None` syntax from 3.10

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+        kucoin_sym = fqsn_to_kucoin_sym(fqsn, self._pairs)
+
+        url = f'/market/candles?type={type}&symbol={kucoin_sym}&startAt={start_dt}&endAt={end_dt}'
+        bars = []
```
type annot here on the list

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+            if not isinstance(data, list):
+                # Do a gradual backoff if Kucoin is rate limiting us
+                backoff_interval = i
```
so this is an `int` from the `range` iter but you're using it as a float to `trio.sleep()` yah?

jaredgoldman replied:
you got it. Very simple logic

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+                    f'History call failed, backing off for {backoff_interval}s')
+                await trio.sleep(backoff_interval)
+            else:
+                bars = data
```
considering you're overriding `bars: list[list]` each backoff algo case, seems weird you pre-define it above as an empty list.

maybe just make the pre-def a type annot only?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+            }
+
+            row = []
+            for _, (field_name, field_type) in enumerate(_ohlc_dtype):
```
hmm, this might be one of the cases where this fancy loop through the hard coded dict you have above might just be better done as a hardcoding of each field in the ohlc array to the respective index in the "bar data" array rx-ed from their api?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+            for _, (field_name, field_type) in enumerate(_ohlc_dtype):
+                value = data[field_name]
+
+                match field_name:
```
yeah i think in general this whole "fill loop" stuff is probably just better done by hardcoding the array fields to our numpy array fields manually, and just link to their docs.

it's some interesting gymnastics tho 😂

jaredgoldman replied:
lol I copied it from @guilledk

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+        async with ctx.open_stream() as stream:
+            async for pattern in stream:
+                await stream.send(await client.search_symbols(pattern))
+                log.info('Kucoin symbol search opened')
```
maybe move this above?

jaredgoldman replied:
Above what

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+    async with open_cached_client('kucoin') as client:
+        log.info('Starting up quote stream')
+        # loop through symbols and sub to feedz
+        for sym in symbols:
```
wait you're making a new ws per sym?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+                    '''
+                    async with trio.open_nursery() as n:
+                        # TODO: cache this task so it's only called once
+                        async def ping_server():
```
cute `trio` work 🏄🏼

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+                # Spawn the ping task here
+                async with open_ping_task(ws) as ws:
+                    tasks = []
```
this seems generally weird to me, the whole `tasks` bit.

i'd say just either add the subs inline without all the `.append()`ing or just generally drop these 2 kinda weird loops.

i got a feeling you can do the subs requests in maybe one send as well?

jaredgoldman replied:
Couldn't figure out the multiple requests in one send but I've factored out this abstraction.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+                typ, quote = await anext(msg_gen)
+
+                while typ != 'trade':
+                    # TODO: use ``anext()`` when it lands in 3.10!
```
you can drop this comment now, also maybe put a comment about how we don't unblock until we get a first "trade quote"

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+
+                task_status.started((init_msgs, quote))
+                feed_is_live.set()
```
hmm iirc you can actually `.started()` sooner if possible. it's only important to wait on a "real trade quote" (as mentioned above) before setting the `feed_is_live: trio.Event()` i think?

jaredgoldman replied:
seems like task started requires a real quote.

jaredgoldman replied:
I could move it earlier in the execution flow and then provide a fake quote? @goodboy thoughts?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+                    await send_chan.send({sym: msg})
+
+
+def make_sub(sym, connect_id, level='l1') -> dict[str, str | bool] | None:
```
yeah imo way too many layers for all this subs stuff.

Just render the subs dicts inline in the `subscribe()` func above.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+
+@trio_async_generator
+async def stream_messages(ws: NoBsWs, sym: str) -> AsyncGenerator[NoBsWs, dict]:
+    timeouts = 0
```
type annots are nice 😉

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+            continue
+
+
+        if msg.get('subject') != None:
```
you don't need the `!= None` check i don't think?

goodboy replied:
i mean unless you expect a `False`-y value to be returned for this?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+                                'type': 'trade',
+                                'price': float(trade_data.price),
+                                'size': float(trade_data.size),
+                                'broker_ts': trade_data.time,
```
why not use the `last_trade_ts` var you just set to avoid yet another attr lookup.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,670 @@
+                    f'difference in time between load and processing {inow - times[-1]}'
+                )
+
+                if (inow - times[-1]) > 60:
```
heh, you really are copying my janky code 😂

goodboy reviewed (changes requested):
@jaredgoldman this is looking super dupe good bud!

I have a few change requests, but i've tested the branch a bit now and it's feeling good!

I think we should maybe also breakout the live trading control / orders stuff into a new issue for follow up. It'd bee good to land this data feed stuff before 😎

goodboy commented:
Looking the code in the ol' vim i'm seeing a lot of these brother:

![screenshot-2023-04-14_19-03-31](https://user-images.githubusercontent.com/291685/232168377-1da797ac-8bfc-4e0f-a506-21b83bd4d3d0.png)

goodboy commented:
Added patch commit to get things working again after your review changes in #498

jaredgoldman commented:
@goodboy I refactored the `subscribe` and `stream_messages` logic to:
- Only use one ws connection and one `ping_task` 
- Be less nested and more readable

lol, the earlier implementation was wildy inefficient spawning a diff ws connection and ping task for each symbol. In addition I've addressed the comments for which I'm not waiting on a reply from you (no rush). We're also getting to the point where we should consider factoring this into a few different files, especially before adding order control.

goodboy commented on piker/brokers/binance.py:
```diff
@@ -551,7 +551,7 @@ async def subscribe(ws: wsproto.WSConnection):
                 typ, quote = await anext(msg_gen)
 
-            task_status.started((init_msgs,  quote))
+            task_status.started((init_msgs, quote))
```
this is just deleting a space yah?

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+            timeframe: float,
+            end_dt: datetime | None = None,
+            start_dt: datetime | None = None,
+        ) -> tuple[np.ndarray, datetime | None, datetime | None]:  # start  # end
```
i would make this a mult-liner, as i do whenever you have this many diff types in one annotation.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+                )
+
+                if (inow - times[-1]) > 60:
+                    await tractor.breakpoint()
```
drop this 😂

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+
+            continue
+
+        if msg.get('subject'):
```
FYI (though not that you have to change it right now) you could actually match against your `msg` and then use object style matching for accessing all the values of the msg if you prefer:

https://jcristharif.com/msgspec/structs.html#pattern-matching

goodboy replied:
ahh just to follow up, so that means you can use the object attributes as both a match condition, **and** an upacker system for binding to local variables in the scope:
https://peps.python.org/pep-0636/#adding-a-ui-matching-objects

jaredgoldman replied:
Although I do see the value of this, I'm not matching for anything more specific than the subject. I think we're fine to leave this for now.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+@acm
+async def subscribe(ws: wsproto.WSConnection, connect_id, sym):
+    # level 2 sub
+    await ws.send_msg({
```
i'm surprised you can't do both of these subs in one msg?

jaredgoldman replied:
Seems like what's even the point the way their API works? https://docs.kucoin.com/#multiplex @goodboy

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+                    # pass back token, and bool, signalling if we're the writer
+                    # and that history has been written
+                    sym: {
+                        'symbol_info': {
```
i'm happy to change this as part of landing #489 but if you wanted to line up with that you can try to as well.

though i guess that won't make it compat with master.. so maybe not?

jaredgoldman replied:
Maybe handle it 489 in that case.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+                log.info('Kucoin symbol search opened')
+
+@acm
+async def open_ping_task(ws: wsproto.WSConnection, ping_interval, connect_id):
```
missing type annots

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+        type: str = '1min',
+    ) -> np.ndarray:
+        '''
+        Get OHLC data and convert to numpy array for perffff:
```
bootyful

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,685 @@
+                api_v='v1',
+            )
+
+            if not isinstance(data, list):
```
what is the normal type of `data` ?
would be handy to type annot this even if it is somewhat variable 😉

goodboy reviewed (changes requested):
Other then this linting output (which i think i've hassled
![screenshot-2023-04-17_18-29-54](https://user-images.githubusercontent.com/291685/232627580-b4be99c5-fa85-444c-abbf-5fd6541e9363.png)
 about enough at this point) 😂 
 
 This is **sooo close**.
 
 I have couple more quiffs mostly to do with typing 🥱 
 
 Testing says this is basically ready to land, even playing with the paper clearing engine a bunch as well as overlays 👍🏼

goodboy commented:
> We're also getting to the point where we should consider factoring this into a few different files, especially before adding order control.

yes, if you want to in this PR it's fine by me. currently we something like:
- core backend api client and general utils in `.api`
- data feed setup, search task and history client in a `.feed`
- all the trade control stuff (which i don't think you have any of yet) in a `.broker`

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,718 @@
+@acm
+async def open_history_client(
+    symbol: str,
+    type: str = '1m',
```
sorry just noticed this..
try not to use built-ins as var names if you can, also i don't think this is used in the func body?

goodboy replied:
yeah i think you can just drop it no?

`timeframe` info is passed into the delivered closure on call.

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,718 @@
+
+            continue
+
+        if msg.get('subject'):
```
you *could* do the match on `msg` directly and just do patterns against the `.subject` field, but honestly not a big deal right now; we can polish it up a bit later.

also, you should normally have a `case _:` and report non-matches that aren't really expected 😉

jaredgoldman replied:
Pattern matching against subject doesn't work for some reason after boxing it with the msgspec struct. I've added a default case with a console.warn log for unexpected messages

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,718 @@
+
+
+@acm
+async def subscribe(ws: wsproto.WSConnection, connect_id, sym):
```
looks wayyy bettter

goodboy commented on piker/brokers/kucoin.py:
```diff
@@ -0,0 +1,718 @@
+                }
+
+                async with (
+                    subscribe(ws, connect_id, kucoin_sym),
```
FYI, you can also pass this into `open_autorecon_ws(ws_url, fixture=subscribe) as ws,` above. Again nbd we can polish it later especially since that API has been reworked entirely in #489 😂

jaredgoldman replied:
Ah but then I wouldn't have my symbol data.

goodboy reviewed (approved):
@jaredgoldman other then a couple more suggestions that I don't think you need to change right now in this PR (particularly bc i just had to entirely rework `open_autiorecon_ws()` in #489 😂) and this little bit of linting i was still able to catch:

![screenshot-2023-04-21_14-11-19](https://user-images.githubusercontent.com/291685/233706584-aa91e7a7-e410-4411-8fb9-7d3e3daa37c5.png)

I think this is more then ready to land!

Great work yet again and glad to see our backend crypto set expanding using `trio` native machnery 🏄🏼 

**Only last thing** is please move all the outstanding todo tasks and follow up into another issue that links to this one **before you merge this**.

Other then that please merge at your leisure 🐶

jaredgoldman commented:
@goodboy https://github.com/pikers/piker/issues/499

goodboy commented:
@jaredgoldman :boom:

excellent work yet again compadre 🏄🏼 

nice follow up issue and final linting before getting this in ma dawg!
